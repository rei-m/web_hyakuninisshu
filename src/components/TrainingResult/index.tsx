import * as React from 'react';
import { Link } from 'react-router-dom';
// import styled from 'styled-components';

export interface TrainingResultOwnProps {
  readonly totalCount: number;
  readonly correctCount: number;
  readonly averageAnswerSecond: number;
}

export interface TrainingResultDispatchProps {
  readonly onClickRestart: () => void;
}

export type TrainingResultProps = TrainingResultOwnProps &
  TrainingResultDispatchProps;

const TrainingResult = ({
  averageAnswerSecond,
  correctCount,
  totalCount,
  onClickRestart
}: TrainingResultProps) => {
  return (
    <section>
      <div>
        <div>正解数</div>
        <div>
          {correctCount} / {totalCount}
        </div>
      </div>
      <div>
        <div>平均回答時間</div>
        <div>{averageAnswerSecond}秒</div>
      </div>
      {correctCount !== totalCount && (
        <button onClick={onClickRestart}>間違えた歌の練習</button>
      )}
      <Link to="/training" replace={true}>
        メニューに戻る
      </Link>
    </section>
  );
};

export default TrainingResult;
