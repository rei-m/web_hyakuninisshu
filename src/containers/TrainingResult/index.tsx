import * as React from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import TrainingResultView, { Props as TrainingResultViewProps } from '@src/components/TrainingResultView';
import ErrorMessage from '@src/components/ErrorMessage';
import { GlobalState } from '@src/state';
import { restartQuestions, QuestionsActions } from '@src/actions/questions';
import { QuestionState } from '@src/enums';
import { Answer } from '@src/types';

export interface ConnectedProps {
  answers: Answer[];
  totalCount: number;
  questionState?: QuestionState;
}

export type DispatchProps = Pick<TrainingResultViewProps, 'onClickRestart'>;

export type Props = ConnectedProps & DispatchProps;

export const TrainingResult: React.FC<Props> = ({ answers, totalCount, questionState, onClickRestart }) => {
  if (questionState !== QuestionState.Finished) {
    return <ErrorMessage text="不正な遷移を行いました。前の画面からやり直してください。" />;
  }
  const averageAnswerSecond = answers.reduce((prev, current) => prev + current.time, 0) / 1000 / totalCount;
  return (
    <TrainingResultView
      averageAnswerSecond={Math.round(averageAnswerSecond * 100) / 100}
      totalCount={totalCount}
      correctCount={answers.filter(a => a.correct).length}
      onClickRestart={onClickRestart}
    />
  );
};

export const mapStateToProps = ({ questions }: GlobalState): ConnectedProps => {
  const { answers, questionState } = questions;
  return {
    answers,
    totalCount: questions.questions.length,
    questionState,
  };
};

export const mapDispatchToProps = (dispatch: ThunkDispatch<GlobalState, {}, QuestionsActions>): DispatchProps => ({
  onClickRestart: () => {
    dispatch(restartQuestions());
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TrainingResult);
