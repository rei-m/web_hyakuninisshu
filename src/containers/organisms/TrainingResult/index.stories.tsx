import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Props, TrainingResult } from './index';
import { create } from '@helper/factory';
import { Answer } from '@src/types';

const answers = [...Array(100).keys()].map(_ => create<Answer>('answer'));

const props: Props = {
  answers,
  onClickBack: action('onClickBack'),
  onClickRestart: action('onClickRestart'),
};

storiesOf('organisms/TrainingResult', module).add('default', () => <TrainingResult {...props} />);
