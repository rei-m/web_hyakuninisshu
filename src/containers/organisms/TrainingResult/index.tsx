import React from 'react';
import { useSelector } from 'react-redux';
import KarutaPlayingResult from '@src/components/organisms/KarutaPlayingResult';
import { GlobalState } from '@src/state';
import { questionsTypes } from '@src/state/questions';
import { Answer } from '@src/types';

export interface OwnProps {
  onClickBack: () => void;
  onClickRestart: () => void;
}

export interface ConnectedProps {
  answers: Answer[];
}

export type PresenterProps = OwnProps & ConnectedProps;

export type ContainerProps = OwnProps & { presenter: React.FC<PresenterProps> };

export const TrainingResultPresenter = ({ answers, onClickBack, onClickRestart }: PresenterProps) => (
  <KarutaPlayingResult answers={answers} onClickBack={onClickBack} onClickRestart={onClickRestart} />
);

export const TrainingResultContainer = ({ presenter, onClickBack, onClickRestart }: ContainerProps) => {
  const { answers } = useSelector<GlobalState, questionsTypes.State>(state => state.questions);
  return presenter({ answers: answers ? answers : [], onClickBack, onClickRestart });
};

export const TrainingResult = (props: OwnProps) => (
  <TrainingResultContainer {...props} presenter={TrainingResultPresenter} />
);

export default TrainingResult;
