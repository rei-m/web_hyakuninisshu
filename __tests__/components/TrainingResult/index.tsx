import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import TrainingResult, { Props } from '@src/components/TrainingResult';
import { mockStaticQuery, sel } from '@test/helpers';

describe('<TrainingResult />', () => {
  let baseProps: Props;

  beforeEach(() => {
    mockStaticQuery.mockImplementationOnce(({ render }) =>
      render({
        trainingResultBGImage: {
          publicURL: '/tatami.jpg',
        },
      })
    );
    baseProps = {
      averageAnswerSecond: 2.5,
      correctCount: 8,
      onClickRestart: jest.fn(),
      totalCount: 10,
    };
  });

  it('should render component when has wrong', () => {
    const renderer = ReactTestRenderer.create(<TrainingResult {...baseProps} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render component when has not wrong', () => {
    const props = { ...baseProps, correctCount: 10 };
    const renderer = ReactTestRenderer.create(<TrainingResult {...props} />);
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should fire onClickRestart when restart clicked', () => {
    const wrapper = shallow(<TrainingResult {...baseProps} />).dive();
    wrapper.find(sel('restart')).simulate('click');
    expect(baseProps.onClickRestart).toHaveBeenCalled();
  });
});
