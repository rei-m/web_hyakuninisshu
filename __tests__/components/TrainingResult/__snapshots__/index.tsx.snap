// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrainingResult /> should render component when has not wrong 1`] = `
<div
  className="sc-bxivhb kvnfBs"
>
  <div
    className="sc-ifAKCX cqqCIC"
  >
    <div
      className="sc-bdVaJa jXieMO"
      style={
        Object {
          "marginBottom": "16px",
        }
      }
    >
      <div
        className="sc-bwzfXH ijMQBI"
      >
        正解数
      </div>
      <div
        className="sc-htpNat hHyKqi"
      >
        10 / 10
      </div>
    </div>
    <div
      className="sc-bdVaJa jXieMO"
      style={
        Object {
          "marginBottom": "32px",
        }
      }
    >
      <div
        className="sc-bwzfXH ijMQBI"
      >
        平均回答時間
      </div>
      <div
        className="sc-htpNat hHyKqi"
      >
        2.5秒
      </div>
    </div>
    <div
      className="sc-EHOje fCNSzj"
    >
      <a
        className="twitter-share-button"
        data-show-count="false"
        data-size="large"
        data-text="百人一首で 10問中 10問 正解しました！"
        data-url="https://hyakuninanki.net"
        href="https://twitter.com/share"
      >
        Tweet
      </a>
    </div>
    <button
      className="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 Component-button-1 Component-normalButton-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-7"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-35 MuiIcon-fontSizeSmall-42 Component-normalIcon-4"
        >
          arrow_back
        </span>
        メニューに戻る
      </span>
      <span
        className="MuiTouchRipple-root-44"
      />
    </button>
  </div>
</div>
`;

exports[`<TrainingResult /> should render component when has wrong 1`] = `
<div
  className="sc-bxivhb kvnfBs"
>
  <div
    className="sc-ifAKCX cqqCIC"
  >
    <div
      className="sc-bdVaJa jXieMO"
      style={
        Object {
          "marginBottom": "16px",
        }
      }
    >
      <div
        className="sc-bwzfXH ijMQBI"
      >
        正解数
      </div>
      <div
        className="sc-htpNat hHyKqi"
      >
        8 / 10
      </div>
    </div>
    <div
      className="sc-bdVaJa jXieMO"
      style={
        Object {
          "marginBottom": "32px",
        }
      }
    >
      <div
        className="sc-bwzfXH ijMQBI"
      >
        平均回答時間
      </div>
      <div
        className="sc-htpNat hHyKqi"
      >
        2.5秒
      </div>
    </div>
    <div
      className="sc-EHOje fCNSzj"
    >
      <a
        className="twitter-share-button"
        data-show-count="false"
        data-size="large"
        data-text="百人一首で 10問中 8問 正解しました！"
        data-url="https://hyakuninanki.net"
        href="https://twitter.com/share"
      >
        Tweet
      </a>
    </div>
    <button
      className="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 Component-button-1 Component-normalButton-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "margin": "16px",
        }
      }
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-7"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-35 MuiIcon-fontSizeSmall-42 Component-normalIcon-4"
        >
          refresh
        </span>
        間違えた歌の練習をする
      </span>
      <span
        className="MuiTouchRipple-root-44"
      />
    </button>
    <button
      className="MuiButtonBase-root-32 MuiButton-root-6 MuiButton-outlined-14 Component-button-1 Component-normalButton-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-7"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-35 MuiIcon-fontSizeSmall-42 Component-normalIcon-4"
        >
          arrow_back
        </span>
        メニューに戻る
      </span>
      <span
        className="MuiTouchRipple-root-44"
      />
    </button>
  </div>
</div>
`;
